{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### *cis*- and *trans*-QTL mapping with tensorQTL\n",
    "\n",
    "This notebook provides examples for running *cis*- and *trans*-QTL mapping with tensorQTL, using open-access data from the [GEUVADIS](https://www.ebi.ac.uk/arrayexpress/experiments/E-GEUV-1/) project.\n",
    "\n",
    "#### Requirements\n",
    "An environment configured with a GPU and ~50GB of memory.\n",
    "\n",
    "#### Test dataset\n",
    "\n",
    "*Note: these files are provided for testing/benchmarking purposes only. They do not constitute an official release from the GEUVADIS project, and no quality-control was applied.*\n",
    "\n",
    "Genotypes in PLINK2 format (chr18 only), and normalized expression data are available [in this repository](./data/); the full dataset is available at [gs://gtex-resources/test_data/geuvadis](https://console.cloud.google.com/storage/browser/gtex-resources/test_data/geuvadis) ([requester pays](https://cloud.google.com/storage/docs/requester-pays))."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "torch: 2.4.0+cu121 (CUDA 12.1), device: cuda\n",
      "pandas 2.2.2\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import torch\n",
    "import tensorqtl\n",
    "from tensorqtl import pgen, cis, trans, post\n",
    "device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")\n",
    "print(f\"torch: {torch.__version__} (CUDA {torch.version.cuda}), device: {device}\")\n",
    "print(f\"pandas {pd.__version__}\")\n",
    "\n",
    "# define paths to data\n",
    "plink_prefix_path = 'data/GEUVADIS.445_samples.GRCh38.20170504.maf01.filtered.nodup.chr18'\n",
    "expression_bed = 'data/GEUVADIS.445_samples.expression.bed.gz'\n",
    "covariates_file = 'data/GEUVADIS.445_samples.covariates.txt'\n",
    "prefix = 'GEUVADIS.445_samples'\n",
    "\n",
    "# load phenotypes and covariates\n",
    "phenotype_df, phenotype_pos_df = tensorqtl.read_phenotype_bed(expression_bed)\n",
    "covariates_df = pd.read_csv(covariates_file, sep='\\t', index_col=0).T\n",
    "\n",
    "# PLINK reader for genotypes\n",
    "pgr = pgen.PgenReader(plink_prefix_path)\n",
    "genotype_df = pgr.load_genotypes()\n",
    "variant_df = pgr.variant_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### *cis*-QTL: nominal p-values for all variant-phenotype pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 445 samples\n",
      "  * 301 phenotypes\n",
      "  * 26 covariates\n",
      "  * 367759 variants\n",
      "  * cis-window: ±1,000,000\n",
      "  * checking phenotypes: 301/301\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 301/301\n",
      "    time elapsed: 0.04 min\n",
      "    * writing output\n",
      "done.\n"
     ]
    }
   ],
   "source": [
    "# map all cis-associations (results for each chromosome are written to file)\n",
    "\n",
    "# all genes\n",
    "# cis.map_nominal(genotype_df, variant_df, phenotype_df, phenotype_pos_df, prefix, covariates_df=covariates_df)\n",
    "\n",
    "# genes on chr18\n",
    "cis.map_nominal(genotype_df, variant_df,\n",
    "                phenotype_df.loc[phenotype_pos_df['chr'] == 'chr18'],\n",
    "                phenotype_pos_df.loc[phenotype_pos_df['chr'] == 'chr18'],\n",
    "                prefix, covariates_df=covariates_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>start_distance</th>\n",
       "      <th>af</th>\n",
       "      <th>ma_samples</th>\n",
       "      <th>ma_count</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>slope</th>\n",
       "      <th>slope_se</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000263006.6</td>\n",
       "      <td>chr18_10644_C_G_b38</td>\n",
       "      <td>-98421</td>\n",
       "      <td>0.016854</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>0.580873</td>\n",
       "      <td>-0.117761</td>\n",
       "      <td>0.213125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000263006.6</td>\n",
       "      <td>chr18_10847_C_A_b38</td>\n",
       "      <td>-98218</td>\n",
       "      <td>0.019101</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>0.142884</td>\n",
       "      <td>-0.298726</td>\n",
       "      <td>0.203505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000263006.6</td>\n",
       "      <td>chr18_11275_G_A_b38</td>\n",
       "      <td>-97790</td>\n",
       "      <td>0.024719</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>0.745231</td>\n",
       "      <td>0.054619</td>\n",
       "      <td>0.167981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000263006.6</td>\n",
       "      <td>chr18_11358_G_A_b38</td>\n",
       "      <td>-97707</td>\n",
       "      <td>0.024719</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>0.745231</td>\n",
       "      <td>0.054619</td>\n",
       "      <td>0.167981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000263006.6</td>\n",
       "      <td>chr18_11445_G_A_b38</td>\n",
       "      <td>-97620</td>\n",
       "      <td>0.023596</td>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>0.603276</td>\n",
       "      <td>0.089378</td>\n",
       "      <td>0.171851</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        phenotype_id           variant_id  start_distance        af  \\\n",
       "0  ENSG00000263006.6  chr18_10644_C_G_b38          -98421  0.016854   \n",
       "1  ENSG00000263006.6  chr18_10847_C_A_b38          -98218  0.019101   \n",
       "2  ENSG00000263006.6  chr18_11275_G_A_b38          -97790  0.024719   \n",
       "3  ENSG00000263006.6  chr18_11358_G_A_b38          -97707  0.024719   \n",
       "4  ENSG00000263006.6  chr18_11445_G_A_b38          -97620  0.023596   \n",
       "\n",
       "   ma_samples  ma_count  pval_nominal     slope  slope_se  \n",
       "0          15        15      0.580873 -0.117761  0.213125  \n",
       "1          17        17      0.142884 -0.298726  0.203505  \n",
       "2          22        22      0.745231  0.054619  0.167981  \n",
       "3          22        22      0.745231  0.054619  0.167981  \n",
       "4          21        21      0.603276  0.089378  0.171851  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load results\n",
    "pairs_df = pd.read_parquet(f'{prefix}.cis_qtl_pairs.chr18.parquet')\n",
    "pairs_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### *cis*-QTL: empirical p-values for phenotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 445 samples\n",
      "  * 301 phenotypes\n",
      "  * 26 covariates\n",
      "  * 367759 variants\n",
      "  * cis-window: ±1,000,000\n",
      "  * using seed 123456\n",
      "  * checking phenotypes: 301/301\n",
      "  * computing permutations\n",
      "    processing phenotype 301/301\n",
      "  Time elapsed: 0.31 min\n",
      "done.\n",
      "Computing q-values\n",
      "  * Number of phenotypes tested: 301\n",
      "  * Correlation between Beta-approximated and empirical p-values: 1.0000\n",
      "  * Calculating q-values with lambda = 0.850\n",
      "  * Proportion of significant phenotypes (1-pi0): 0.76\n",
      "  * QTL phenotypes @ FDR 0.05: 205\n",
      "  * min p-value threshold @ FDR 0.05: 0.135284\n"
     ]
    }
   ],
   "source": [
    "# all genes\n",
    "# cis_df = cis.map_cis(genotype_df, variant_df, phenotype_df, phenotype_pos_df, covariates_df=covariates_df)\n",
    "\n",
    "# genes on chr18\n",
    "cis_df = cis.map_cis(genotype_df, variant_df, \n",
    "                     phenotype_df.loc[phenotype_pos_df['chr'] == 'chr18'],\n",
    "                     phenotype_pos_df.loc[phenotype_pos_df['chr'] == 'chr18'],\n",
    "                     covariates_df=covariates_df, seed=123456)\n",
    "# compute q-values (in practice, this must be run on all genes, not a subset)\n",
    "post.calculate_qvalues(cis_df, fdr=0.05, qvalue_lambda=0.85)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>num_var</th>\n",
       "      <th>beta_shape1</th>\n",
       "      <th>beta_shape2</th>\n",
       "      <th>true_df</th>\n",
       "      <th>pval_true_df</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>start_distance</th>\n",
       "      <th>end_distance</th>\n",
       "      <th>ma_samples</th>\n",
       "      <th>ma_count</th>\n",
       "      <th>af</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>slope</th>\n",
       "      <th>slope_se</th>\n",
       "      <th>pval_perm</th>\n",
       "      <th>pval_beta</th>\n",
       "      <th>qval</th>\n",
       "      <th>pval_nominal_threshold</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>phenotype_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000263006.6</th>\n",
       "      <td>6120</td>\n",
       "      <td>1.038811</td>\n",
       "      <td>1138.434082</td>\n",
       "      <td>374.660400</td>\n",
       "      <td>8.220950e-40</td>\n",
       "      <td>chr18_112535_G_A_b38</td>\n",
       "      <td>3470</td>\n",
       "      <td>3470</td>\n",
       "      <td>212</td>\n",
       "      <td>251</td>\n",
       "      <td>0.282022</td>\n",
       "      <td>4.050344e-44</td>\n",
       "      <td>0.726425</td>\n",
       "      <td>0.046171</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>3.677735e-38</td>\n",
       "      <td>2.697006e-37</td>\n",
       "      <td>0.000141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000101557.14</th>\n",
       "      <td>6355</td>\n",
       "      <td>1.032237</td>\n",
       "      <td>1076.303223</td>\n",
       "      <td>370.176422</td>\n",
       "      <td>5.632806e-11</td>\n",
       "      <td>chr18_210698_T_C_b38</td>\n",
       "      <td>52315</td>\n",
       "      <td>52315</td>\n",
       "      <td>192</td>\n",
       "      <td>222</td>\n",
       "      <td>0.249438</td>\n",
       "      <td>3.505411e-12</td>\n",
       "      <td>-0.191712</td>\n",
       "      <td>0.026749</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>3.498951e-08</td>\n",
       "      <td>3.563747e-08</td>\n",
       "      <td>0.000146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000079134.11</th>\n",
       "      <td>6921</td>\n",
       "      <td>1.047219</td>\n",
       "      <td>1155.660156</td>\n",
       "      <td>370.356049</td>\n",
       "      <td>3.888738e-08</td>\n",
       "      <td>chr18_243547_T_A_b38</td>\n",
       "      <td>-24503</td>\n",
       "      <td>-24503</td>\n",
       "      <td>293</td>\n",
       "      <td>383</td>\n",
       "      <td>0.430337</td>\n",
       "      <td>5.473709e-09</td>\n",
       "      <td>-0.122720</td>\n",
       "      <td>0.020602</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>2.743975e-05</td>\n",
       "      <td>1.916427e-05</td>\n",
       "      <td>0.000141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000263884.1</th>\n",
       "      <td>6921</td>\n",
       "      <td>1.039806</td>\n",
       "      <td>1152.501587</td>\n",
       "      <td>369.873505</td>\n",
       "      <td>7.681884e-04</td>\n",
       "      <td>chr18_584440_G_C_b38</td>\n",
       "      <td>316292</td>\n",
       "      <td>316292</td>\n",
       "      <td>81</td>\n",
       "      <td>88</td>\n",
       "      <td>0.098876</td>\n",
       "      <td>3.540399e-04</td>\n",
       "      <td>-0.330811</td>\n",
       "      <td>0.091845</td>\n",
       "      <td>0.574843</td>\n",
       "      <td>5.695498e-01</td>\n",
       "      <td>1.577698e-01</td>\n",
       "      <td>0.000139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000158270.11</th>\n",
       "      <td>8134</td>\n",
       "      <td>1.054919</td>\n",
       "      <td>1277.927246</td>\n",
       "      <td>369.469086</td>\n",
       "      <td>2.516529e-09</td>\n",
       "      <td>chr18_519222_C_T_b38</td>\n",
       "      <td>18500</td>\n",
       "      <td>18500</td>\n",
       "      <td>108</td>\n",
       "      <td>115</td>\n",
       "      <td>0.129213</td>\n",
       "      <td>2.409717e-10</td>\n",
       "      <td>-0.388277</td>\n",
       "      <td>0.059808</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>1.567348e-06</td>\n",
       "      <td>1.321136e-06</td>\n",
       "      <td>0.000130</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    num_var  beta_shape1  beta_shape2     true_df  \\\n",
       "phenotype_id                                                        \n",
       "ENSG00000263006.6      6120     1.038811  1138.434082  374.660400   \n",
       "ENSG00000101557.14     6355     1.032237  1076.303223  370.176422   \n",
       "ENSG00000079134.11     6921     1.047219  1155.660156  370.356049   \n",
       "ENSG00000263884.1      6921     1.039806  1152.501587  369.873505   \n",
       "ENSG00000158270.11     8134     1.054919  1277.927246  369.469086   \n",
       "\n",
       "                    pval_true_df            variant_id  start_distance  \\\n",
       "phenotype_id                                                             \n",
       "ENSG00000263006.6   8.220950e-40  chr18_112535_G_A_b38            3470   \n",
       "ENSG00000101557.14  5.632806e-11  chr18_210698_T_C_b38           52315   \n",
       "ENSG00000079134.11  3.888738e-08  chr18_243547_T_A_b38          -24503   \n",
       "ENSG00000263884.1   7.681884e-04  chr18_584440_G_C_b38          316292   \n",
       "ENSG00000158270.11  2.516529e-09  chr18_519222_C_T_b38           18500   \n",
       "\n",
       "                    end_distance  ma_samples  ma_count        af  \\\n",
       "phenotype_id                                                       \n",
       "ENSG00000263006.6           3470         212       251  0.282022   \n",
       "ENSG00000101557.14         52315         192       222  0.249438   \n",
       "ENSG00000079134.11        -24503         293       383  0.430337   \n",
       "ENSG00000263884.1         316292          81        88  0.098876   \n",
       "ENSG00000158270.11         18500         108       115  0.129213   \n",
       "\n",
       "                    pval_nominal     slope  slope_se  pval_perm     pval_beta  \\\n",
       "phenotype_id                                                                    \n",
       "ENSG00000263006.6   4.050344e-44  0.726425  0.046171   0.000100  3.677735e-38   \n",
       "ENSG00000101557.14  3.505411e-12 -0.191712  0.026749   0.000100  3.498951e-08   \n",
       "ENSG00000079134.11  5.473709e-09 -0.122720  0.020602   0.000100  2.743975e-05   \n",
       "ENSG00000263884.1   3.540399e-04 -0.330811  0.091845   0.574843  5.695498e-01   \n",
       "ENSG00000158270.11  2.409717e-10 -0.388277  0.059808   0.000100  1.567348e-06   \n",
       "\n",
       "                            qval  pval_nominal_threshold  \n",
       "phenotype_id                                              \n",
       "ENSG00000263006.6   2.697006e-37                0.000141  \n",
       "ENSG00000101557.14  3.563747e-08                0.000146  \n",
       "ENSG00000079134.11  1.916427e-05                0.000141  \n",
       "ENSG00000263884.1   1.577698e-01                0.000139  \n",
       "ENSG00000158270.11  1.321136e-06                0.000130  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cis_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### *trans*-QTL mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "trans-QTL mapping\n",
      "  * 445 samples\n",
      "  * 19836 phenotypes\n",
      "  * 26 covariates\n",
      "  * 367759 variants\n",
      "    processing batch 37/37\n",
      "    elapsed time: 0.02 min\n",
      "  * 210838 variants passed MAF >= 0.05 filtering\n",
      "done.\n"
     ]
    }
   ],
   "source": [
    "# run mapping\n",
    "# to limit output size, only associations with p-value <= 1e-5 are returned\n",
    "trans_df = trans.map_trans(genotype_df, phenotype_df, covariates_df, batch_size=10000,\n",
    "                           return_sparse=True, pval_threshold=1e-5, maf_threshold=0.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove cis-associations\n",
    "trans_df = trans.filter_cis(trans_df, phenotype_pos_df, variant_df, window=5000000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>pval</th>\n",
       "      <th>b</th>\n",
       "      <th>b_se</th>\n",
       "      <th>af</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr18_20683_A_G_b38</td>\n",
       "      <td>ENSG00000163900.10</td>\n",
       "      <td>5.012229e-06</td>\n",
       "      <td>0.209540</td>\n",
       "      <td>0.045309</td>\n",
       "      <td>0.179775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr18_27346_G_T_b38</td>\n",
       "      <td>ENSG00000164088.17</td>\n",
       "      <td>7.309937e-06</td>\n",
       "      <td>-0.265623</td>\n",
       "      <td>0.058483</td>\n",
       "      <td>0.123596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr18_43564_G_A_b38</td>\n",
       "      <td>ENSG00000198162.12</td>\n",
       "      <td>1.314060e-07</td>\n",
       "      <td>-0.202922</td>\n",
       "      <td>0.037792</td>\n",
       "      <td>0.093258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr18_43564_G_A_b38</td>\n",
       "      <td>ENSG00000261098.1</td>\n",
       "      <td>8.494569e-06</td>\n",
       "      <td>-0.421968</td>\n",
       "      <td>0.093594</td>\n",
       "      <td>0.093258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr18_43611_C_T_b38</td>\n",
       "      <td>ENSG00000265972.5</td>\n",
       "      <td>1.448981e-06</td>\n",
       "      <td>-0.272301</td>\n",
       "      <td>0.055697</td>\n",
       "      <td>0.135955</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             variant_id        phenotype_id          pval         b      b_se  \\\n",
       "1   chr18_20683_A_G_b38  ENSG00000163900.10  5.012229e-06  0.209540  0.045309   \n",
       "3   chr18_27346_G_T_b38  ENSG00000164088.17  7.309937e-06 -0.265623  0.058483   \n",
       "11  chr18_43564_G_A_b38  ENSG00000198162.12  1.314060e-07 -0.202922  0.037792   \n",
       "12  chr18_43564_G_A_b38   ENSG00000261098.1  8.494569e-06 -0.421968  0.093594   \n",
       "13  chr18_43611_C_T_b38   ENSG00000265972.5  1.448981e-06 -0.272301  0.055697   \n",
       "\n",
       "          af  \n",
       "1   0.179775  \n",
       "3   0.123596  \n",
       "11  0.093258  \n",
       "12  0.093258  \n",
       "13  0.135955  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trans_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
